apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -530,297 +533,605 @@ public Flow(long maxByteCount, long maxElementCount, Action<Task> registerTaskFn
+                List<T> result = new List<T>();
+                Queue<T> q = _qs.First?.Value;
+                if (q != null)
+                {
+                    while (result.Count < maxCount)
+                    {
+                        if (q.Count > 0)
+                        {
+                            var item = q.Dequeue();
+                            _requeueCount -= 1;
+                            if (includeFn == null || includeFn(item))
+                            {
+                                result.Add(item);
+                            }
+                        }
+                        else
+                        {
+                            _qs.RemoveFirst();
+                            q = _qs.First?.Value;
+                            if (q == null)
+                            {
+                                break;
+                            }
+                        }
+                    }
+                }
+                while (result.Count < maxCount && _q.Count > 0)
+                {
+                    var item = _q.Dequeue();
+                    if (includeFn == null || includeFn(item))
+                    {
+                        result.Add(item);
+                    }
+                }
+                return result;
+            }
+
+            public bool TryPeek(out T value)
+            {
+                var qsNode = _qs.First;
+                while (qsNode != null)
+                {
+                    if (qsNode.Value.Count > 0)
+                    {
+                        value = qsNode.Value.Peek();
+                        return true;
+                    }
+                    qsNode = qsNode.Next;
+                }
+                if (_q.Count > 0)
+                {
+                    value = _q.Peek();
+                    return true;
+                }
+                value = default(T);
+                return false;
+            }
+        }
+
+        internal class AsyncSingleRecvQueue<T>
+        {
+            public AsyncSingleRecvQueue(TaskHelper taskHelper) => _taskHelper = taskHelper;
+
+            private readonly TaskHelper _taskHelper;
+            private readonly object _lock = new object();
+            private readonly Queue<T> _queue = new Queue<T>();
+            private TaskCompletionSource<int> _tcs = null;
+
+            // Thread-safe.
+            public void Enqueue(T item)
+            {
+                TaskCompletionSource<int> tcs;
+                lock (_lock)
+                {
+                    _queue.Enqueue(item);
+                    tcs = _tcs;
+                }
+                if (tcs != null)
+                {
+                    // Don't run in lock, as it may execute continuations synchonously.
+                    tcs.SetResult(0);
+                }
+            }
+
+            // Thread-safe, but only one dequeue is allowed at a time.
+            public async Task<T> DequeueAsync()
```

Comment:
I don't think this is safe. If two threads call DequeueAsync at the same time, they could each create and assign a separate TCS, one of which would get lost. (And an Enqueue call at the same time could set the result on the first one, so potentially neither thread sees the new item!). Happy to VC to explain what I mean. Perhaps ConcurrentQueue is useful here?